counter: enable kinetis pit counter driver

hakehuang opened the pull request:
enable kinetis pit counter driver on frdm_k82f
tested with counter_basic_api case.
this PR depends on https://github.com/zephyrproject-rtos/hal_nxp/pull/47

MaureenHelm commented on dts/bindings/rtc/nxp,kinetis-pit.yaml:
```diff
@@ -0,0 +1,18 @@
+
+compatible: "nxp,kinetis-pit"
+
+include: [base.yaml]
```
Include `rtc.yaml` instead

hakehuang replied:
ok.

MaureenHelm commented on boards/arm/frdm_k82f/frdm_k82f.yaml:
```diff
@@ -14,6 +14,7 @@ supported:
   - gpio
   - i2c
   - nvs
+  - pit
```
Replace with `counter`

hakehuang replied:
ok

MaureenHelm replied:
I don't think it's a problem if there are multiple counters. This is a list of supported interfaces, not specific IPs.

hakehuang replied:
ok

MaureenHelm commented on boards/arm/frdm_k82f/frdm_k82f_defconfig:
```diff
@@ -11,6 +11,7 @@ CONFIG_PINMUX=y
 CONFIG_GPIO=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=120000000
 CONFIG_OSC_LOW_POWER=y
+CONFIG_COUNTER=y
```
This will enable the counter in every application for this board, including blinky and hello_world. Please remove.

hakehuang replied:
ok

MaureenHelm commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -0,0 +1,209 @@
+	return 0;
+}
+
+static int mcux_pit_stop(struct device *dev)
```
Hehe, I love this function name. :)
https://www.merriam-webster.com/dictionary/pit%20stop

MaureenHelm commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -0,0 +1,209 @@
+		.channels = 0,
+	},
+	.base = (PIT_Type *)DT_INST_REG_ADDR(0),
+	.pit_channel = CONFIG_PIT_CHANNEL,
```
This should come from device tree, not Kconfig

hakehuang replied:
there are 4 pit channels, user can select either of them, if not use config interface use need change the device every time.

MaureenHelm replied:
Yes, but it is an instance-specific property which means it should be defined in the device tree.

hakehuang replied:
ok

MaureenHelm commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -0,0 +1,209 @@
+
+/*
+ * This driver is single-instance. If the devicetree contains multiple
+ * instances, this will fail and the driver needs to be revisited.
```
Why not support multiple instances?

hakehuang replied:
in all MCU platform, we only have one PIT, I never test multiply PIT. maybe they are OK, but I copy it from lptmr and not sure I need remove it or not.

MaureenHelm commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -66,9 +66,14 @@ static const char * const devices[] = {
 	LABELS_FOR_DT_COMPAT(st_stm32_rtc)
+#ifdef CONFIG_COUNTER_MCUX_PIT
+	LABELS_FOR_DT_COMPAT(nxp_kinetis_pit)
+#endif
```
Neither of these ifdefs should be necessary

hakehuang replied:
the LABELS_FOR_DT_COMPAT can't tell whether the counter driver for rtc or pit is enable. it only checks whether the RTC0/PIT0 exist or not. and I find if I enable RTC0 in counter test, it fails, I need report one jira issue for this, as it is not default enabled. so we miss this test before.

MaureenHelm replied:
Ok, please create a github issue and copy a link here.

hakehuang replied:
#26832 reported for frdm_k82f RTC counter

MaureenHelm commented on soc/arm/nxp_kinetis/k8x/Kconfig.soc:
```diff
@@ -69,4 +69,6 @@ config K8X_FLASH_CLOCK_DIVIDER
 	  system clock.
 
+config COUNTER_MCUX_PIT
+	default y if HAS_MCUX_PIT
```
Please move these changes to `soc/arm/nxp_kinetis/k8x/Kconfig.defconfig.series`

hakehuang replied:
ok

hakehuang committed:
dts: add pit as counter

add dts for kinetis pit counter driver

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
counter: add pit in k8s board support

enable frdm_k82f with pit counter driver
this driver can be used as period hardware trigger

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
counter: enable pit support for k8x series

enable PIT in k8x series

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
counter: add pit counter driver

add Kinetis PIT as counter driver

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
test: update counter test to support pit

update counter_basic_api to support kinetis PIT
update counter to support multi coutner driver

select PERIOD_TO_US by device name at runtime

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

MaureenHelm commented on west.yml:
```diff
@@ -98,7 +98,8 @@ manifest:
     - name: hal_nxp
-      revision: 478f03a71256aeb2890e685ebf5a1dc6832a6b5f
+      # revision: 478f03a71256aeb2890e685ebf5a1dc6832a6b5f
+      revision: pull/47/head
```
The hal driver has been merged, so you can now update this to the new SHA

hakehuang replied:
ok

MaureenHelm commented on dts/bindings/rtc/nxp,kinetis-pit.yaml:
```diff
@@ -0,0 +1,18 @@
+    reg:
+      required: true
+
+    interrupts:
+      required: true
+
+    label:
+      required: true
```
These properties are required in `rtc.yaml`, so it's redundant to put them here too. You can remove these lines.

hakehuang replied:
ok

MaureenHelm commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -0,0 +1,262 @@
+
+	config->irq_config_func(dev);
+
+	/*hard code to 1s*/
```
Why? This should probably be a dts property in `dts/bindings/rtc/nxp,kinetis-pit.yaml`

hakehuang replied:
ok

MaureenHelm commented on tests/drivers/counter/counter_basic_api/boards/frdm_k82f.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_COUNTER=y
+# CONFIG_COUNTER_MCUX_RTC=y
```
`tests/drivers/counter/counter_basic_api/prj.conf` already sets `CONFIG_COUNTER=y`. You can remove this file

hakehuang replied:
ok

hakehuang commented:
https://github.com/zephyrproject-rtos/zephyr/pull/26629#discussion_r452847247
#26832 reported for frdm_k82f RTC counter

MaureenHelm commented on dts/arm/nxp/nxp_k8x.dtsi:
```diff
@@ -351,7 +351,7 @@
 			reg = <0x400ac000 0x1000>;
 			interrupts = <65 3>;
-			clocks = <&sim KINETIS_SIM_BUS_CLK 0x1030 12>;
+			clocks = <&sim KINETIS_SIM_BUS_CLK 0x103c 23>;
```
Why are you changing a SPI clock gate in a PIT commit? Is this a mistake?

hakehuang replied:
nice shot, this should change to the pit0, I wrongly put it to dspi, although we do not access the clock gate register bit in code.

MaureenHelm commented on drivers/counter/Kconfig.mcux_pit:
```diff
@@ -0,0 +1,10 @@
+# MCUXpresso SDK Periodic Interrupt Timer (PIT)
+
+# Copyright (c) 2020 NXP Semiconductor INC.
```
```suggestion
# Copyright 2020 NXP
```

hakehuang replied:
ok

MaureenHelm commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) 2020 NXP Semiconductor INC.
```
```suggestion
 * Copyright 2020 NXP
```

hakehuang replied:
ok

MaureenHelm commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -342,7 +347,6 @@ void test_single_shot_alarm_instance(const char *dev_name, bool set_top)
 			dev_name, err);
 
 	k_busy_wait(2*(uint32_t)counter_ticks_to_us(dev, ticks));
-
```
spurious change

hakehuang replied:
remove it

jensschroer commented:
Hi, I was just wondering if it is possible to also enable this on the K64f or would this require additional changes?
Thank you!

hakehuang commented:
> Hi, I was just wondering if it is possible to also enable this on the K64f or would this require additional changes?
> Thank you!

@jensschroer , frdmk64 pit can be enabled easily with device tree, no other code change. but as frdmk64f has a counter driver already(RTC), and there is bug pending #26832, or you need disable the rtc counter driver to pass the counter_basic_api test.

hakehuang committed:
dts: enable pit0 in k6s device tree

enable pit0 in NXP k6s

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
counter: enable k6s pit support support

enable k6s pit counter support

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
counter: enable pit counter driver on frdmk64f

enable pit counter driver in frdmk64f

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

MaureenHelm commented on dts/bindings/rtc/nxp,kinetis-pit.yaml:
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2020 NXP Semiconductor INC.
```
```suggestion
# Copyright 2020 NXP
```

hakehuang replied:
ok

MaureenHelm commented on drivers/counter/counter_mcux_pit.c:
```diff
@@ -0,0 +1,259 @@
+
+static void mcux_pit_irq_config_0(struct device *dev);
+
+static struct mcux_pit_config mcux_pit_config_0 = {
```
add `const`

hakehuang replied:
ok

MaureenHelm replied:
That means we have to hard code it in dts by adding a `clock-frequency` property to `dts/bindings/rtc/nxp,kinetis-pit.yaml`. The microchip rtc driver does this.

cc: @nordic-krch @pabigot

hakehuang replied:
ok

MaureenHelm reviewed (changes requested):
Just two nits, but otherwise looks good. Thanks @hakehuang

MaureenHelm commented on soc/arm/nxp_kinetis/k8x/Kconfig.defconfig.series:
```diff
@@ -76,6 +76,9 @@ config WDT_MCUX_WDOG
 	depends on WATCHDOG
 
+config COUNTER_MCUX_PIT
+	default y if HAS_MCUX_PIT
```
```suggestion
	default y if HAS_MCUX_PIT
	depends on COUNTER
```

hakehuang replied:
ok

MaureenHelm commented on soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.series:
```diff
@@ -19,6 +19,9 @@ config DMA_MCUX_EDMA
 	depends on DMA
 
+config COUNTER_MCUX_PIT
+	default y if HAS_MCUX_PIT
```
```suggestion
	default y if HAS_MCUX_PIT
	depends on COUNTER
```

hakehuang replied:
ok

dleach02 commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -174,7 +181,12 @@ void test_set_top_value_with_alarm_instance(const char *dev_name)
+	if (strcmp(dev_name, "RTC_0") == 0) {
+		counter_period_us = COUNTER_PERIOD_US_VAL;
+	} else {
+		counter_period_us = 20000;
```
tough seeing a hard coded value like this in source.. maybe a comment on why for a future viewer of this code?

hakehuang replied:
ok

hakehuang replied:
this value is the default test range for pit to count, RTC drivers need calculated by (USEC_PER_SEC * 2U), so if one platform has two counter driver RTC and PIT then, this will be conflicts, so I separarted them by the dev_name.

dleach02 commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -301,14 +319,20 @@ void test_single_shot_alarm_instance(const char *dev_name, bool set_top)
+	if (strcmp(dev_name, "RTC_0") == 0) {
+		counter_period_us = COUNTER_PERIOD_US_VAL;
+	} else {
+		counter_period_us = 20000;
```
same comment as earlier

hakehuang replied:
ok

dleach02 commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -431,14 +454,20 @@ void test_multiple_alarms_instance(const char *dev_name)
+	if (strcmp(dev_name, "RTC_0") == 0) {
+		counter_period_us = COUNTER_PERIOD_US_VAL;
+	} else {
+		counter_period_us = 20000;
```
ditto

hakehuang replied:
ok

dleach02 commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -517,9 +546,15 @@ void test_all_channels_instance(const char *dev_name)
+	if (strcmp(dev_name, "RTC_0") == 0) {
+		counter_period_us = COUNTER_PERIOD_US_VAL;
+	} else {
+		counter_period_us = 20000;
```
ditto

hakehuang replied:
ok

dleach02 reviewed (changes requested):
Looks good but I think you need to add a comment on the hard coded numbers for future viewers/maintainers of this code.

hakehuang commented:
@dleach02 please help to check my comments in code
